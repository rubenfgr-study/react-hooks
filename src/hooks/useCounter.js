import { useState } from "react";

export default function useCounter(initialState) {
  const [counter, setState] = useState(initialState);

  const increment = (factor = 1) => {
    setState(counter + factor);
  };

  const decrement = (factor = 1) => {
    setState(counter - factor);
  };

  const reset = () => setState(initialState);

  return { counter, increment, decrement, reset };
}
