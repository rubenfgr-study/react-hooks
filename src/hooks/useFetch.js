import { useEffect, useRef, useState } from "react";

export default function useFetch(url) {
  const isMounted = useRef(true);

  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null,
  });

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    setState({
      data: null,
      loading: true,
      error: null,
    });

    fetch(url)
      .then((response) => response.json())
      .then(
        (data) =>
          !!isMounted.current &&
          setState({
            data,
            loading: false,
            error: null,
          })
      );
  }, [url]);

  return state;
}
