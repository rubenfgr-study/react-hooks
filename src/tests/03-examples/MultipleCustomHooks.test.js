import { shallow } from "enzyme";
import MultipleCustomHooks from "../../components/03-examples/MultipleCustomHooks";
import useCounter from "../../hooks/useCounter";
import useFetch from "../../hooks/useFetch";
jest.mock("../../hooks/useFetch");
jest.mock("../../hooks/useCounter");

describe("MultipleCustomHooks Component", () => {
  beforeEach(() => {
    useCounter.mockReturnValue({
      counter: 10,
      increment: () => {},
    });
  });
  test("snapshot", () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test("should be show info", () => {
    useFetch.mockReturnValue({
      data: [{ author: "Walter White", quote: "I am the one who knocks" }],
      loading: false,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper.find(".alert").exists()).toBeFalsy();
    expect(wrapper.find("h1").text()).toBe("BreakingBad Quotes");
    expect(wrapper.find("p").text()).toBe("I am the one who knocks");
    wrapper.find("button").simulate("click");
  });
});
