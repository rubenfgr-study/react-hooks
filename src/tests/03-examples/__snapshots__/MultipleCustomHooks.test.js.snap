// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleCustomHooks Component snapshot 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente Quote
  </button>
</div>
`;
